### PR TITLE
Bindgen and the Beginnings of Rex Bindings

### PR DESCRIPTION
**Pull request by cwfitzgerald**
This is a bit of a beast of a PR, not because it's long, but because it contains a small amount of really complex unsafe code.

The things to review closely are:
- Everything in bve-rex
- The rust code in `bve-rex-sys`
- bve-rex-sys/wrapper/game.{c,h}pp

One of the biggest things I'm worried about is not properly dropping things in the right place. I have added comments where it _should_ be dropped.

### CODE REVIEW

**Comment by cwfitzgerald** [comment]
So that took way too damn long, but it's time for another review as the old one is 20 commits stale. Rex is now a shared library but runs everywhere.

File: bve-rex/src/game/mod.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        // It will be dropped by the `drop_game` function.
+        Box::into_raw(built) as *mut c_void,
+        Some(on_init_wrapper::<B::Built>),
+        Some(on_slice_wrapper::<B::Built>),
+        Some(on_resize_wrapper::<B::Built>),
+        Some(drop_game::<B::Built>),
+    );
+    // Box the game wrapper up and turn it into a pointer.
+    // This is dropped in the `destroy` hook provided to rex.
+    Box::into_raw(Box::new(game))
```
Comment (DethRaid) [inline bve-rex/src/game/mod.rs]: Why do you make a Box from the game pointer, only to immediately un-box it?

File: bve-rex-sys/build.rs [hunk 2]
```diff
@@ -104,48 +74,66 @@ fn match_c_files(files: FileTypes) -> impl Fn(Result<DirEntry, Error>) -> Option
...
 fn main() {
-    #[cfg(target_env = "mingw")]
-    {
-        panic!("Rex does not support mingw");
-    }
-    #[cfg(target_os = "macos")]
-    {
-        panic!("Rex does not support macos");
+    cfg_if::cfg_if! {
+        if #[cfg(target_env = "mingw")] {
```
[the remark this answers is not in the document]
Comment (cwfitzgerald) [reply bve-rex-sys/build.rs]: He did, which is why I clarified that it is `bve-rex-sys` that doesn't support mingw, not rex.

[the remark this answers is not in the document]
Comment (DethRaid) [reply bve-rex-sys/build.rs]: oh

File: bve-rex/src/game/mod.rs (hunk 1 above)
Comment (cwfitzgerald) [reply bve-rex/src/game/mod.rs]: that's the translation from `bve::Game` -> `Box<bve::game>` to `*mut bve::game`

Comment (DethRaid) [reply bve-rex/src/game/mod.rs]: Update: I get it

**Pull request merged**